### Bank Customer churn prediction

Given a Bank customer, can we build a classifier which can determine whether they will leave or not using Neural networks?

### Load tensorflow

In [0]:
import tensorflow as tf

tf.set_random_seed(42)

In [3]:
tf.__version__

'1.15.0-rc3'

### Collect Data

In [4]:
import keras


Using TensorFlow backend.


In [0]:
#Read data from file
import pandas as pd
bank_df = pd.read_csv('./bank.csv',encoding='utf-8')

In [6]:
bank_df.head()

RowNumber  CustomerId   Surname  ...  IsActiveMember EstimatedSalary Exited
0          1    15634602  Hargrave  ...               1       101348.88      1
1          2    15647311      Hill  ...               1       112542.58      0
2          3    15619304      Onio  ...               0       113931.57      1
3          4    15701354      Boni  ...               0        93826.63      0
4          5    15737888  Mitchell  ...               1        79084.10      0

[5 rows x 14 columns]

In [7]:
bank_df.describe().T

count          mean  ...           75%          max
RowNumber        10000.0  5.000500e+03  ...  7.500250e+03     10000.00
CustomerId       10000.0  1.569094e+07  ...  1.575323e+07  15815690.00
CreditScore      10000.0  6.505288e+02  ...  7.180000e+02       850.00
Age              10000.0  3.892180e+01  ...  4.400000e+01        92.00
Tenure           10000.0  5.012800e+00  ...  7.000000e+00        10.00
Balance          10000.0  7.648589e+04  ...  1.276442e+05    250898.09
NumOfProducts    10000.0  1.530200e+00  ...  2.000000e+00         4.00
HasCrCard        10000.0  7.055000e-01  ...  1.000000e+00         1.00
IsActiveMember   10000.0  5.151000e-01  ...  1.000000e+00         1.00
EstimatedSalary  10000.0  1.000902e+05  ...  1.493882e+05    199992.48
Exited           10000.0  2.037000e-01  ...  0.000000e+00         1.00

[11 rows x 8 columns]

In [0]:
#RowNumber and CustomerId are unique columns which may not add value to the information
#Droping these columns
bank_df=bank_df.drop(['RowNumber','CustomerId','Surname'],axis=1)

In [9]:
bank_df.shape

(10000, 11)

In [0]:
#Target column is Exited, Either the customer leave the bank or will not leave
X=bank_df.drop(['Exited'],axis=1)
y=bank_df['Exited']

In [11]:
X.shape

(10000, 10)

In [12]:
X.head()

CreditScore Geography  Gender  ...  HasCrCard  IsActiveMember  EstimatedSalary
0          619    France  Female  ...          1               1        101348.88
1          608     Spain  Female  ...          0               1        112542.58
2          502    France  Female  ...          1               0        113931.57
3          699    France  Female  ...          0               0         93826.63
4          850     Spain  Female  ...          1               1         79084.10

[5 rows x 10 columns]

In [0]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

le=LabelEncoder()
X['Gender'] = le.fit_transform(X.Gender.values)


In [0]:
le=LabelEncoder()
X['Geography'] = le.fit_transform(X.Geography.values)

In [15]:
X.shape

(10000, 10)

In [16]:
oh=OneHotEncoder(categorical_features=[1])
X=oh.fit_transform(X).toarray()

/usr/local/lib/python2.7/dist-packages/sklearn/preprocessing/_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/usr/local/lib/python2.7/dist-packages/sklearn/preprocessing/_encoders.py:392: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [0]:
X=X[:,1:]

In [18]:
X

array([[0.0000000e+00, 0.0000000e+00, 6.1900000e+02, ..., 1.0000000e+00,
        1.0000000e+00, 1.0134888e+05],
       [0.0000000e+00, 1.0000000e+00, 6.0800000e+02, ..., 0.0000000e+00,
        1.0000000e+00, 1.1254258e+05],
       [0.0000000e+00, 0.0000000e+00, 5.0200000e+02, ..., 1.0000000e+00,
        0.0000000e+00, 1.1393157e+05],
       ...,
       [0.0000000e+00, 0.0000000e+00, 7.0900000e+02, ..., 0.0000000e+00,
        1.0000000e+00, 4.2085580e+04],
       [1.0000000e+00, 0.0000000e+00, 7.7200000e+02, ..., 1.0000000e+00,
        0.0000000e+00, 9.2888520e+04],
       [0.0000000e+00, 0.0000000e+00, 7.9200000e+02, ..., 1.0000000e+00,
        0.0000000e+00, 3.8190780e+04]])

In [0]:
#Splitting the data into train and test
import numpy as np
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.24, random_state=42)

In [0]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train=sc.fit_transform(X_train)
X_test=sc.fit_transform(X_test)

In [0]:
from keras.models import Sequential
from keras.layers import Dense,Dropout

In [22]:
model =Sequential()
model.add(Dense(output_dim=6,init="uniform",activation='relu',input_dim=11))
model.add(Dropout(rate=0.1))

model.add(Dense(output_dim=6,init="uniform",activation='relu',input_dim=11))
model.add(Dropout(rate=0.1))

model.add(Dense(output_dim=1,init="uniform",activation='sigmoid'))

W1013 13:40:47.129539 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", activation="relu", input_dim=11, units=6)`
  
W1013 13:40:47.132582 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W1013 13:40:47.145252 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.

W1013 13:40:47.161349 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dis

In [23]:
#Compile the model - add Loss and Gradient Descent optimizer
model.compile(optimizer='adam', loss='binary_crossentropy',metrics=['accuracy'])

W1013 13:40:55.214215 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

W1013 13:40:55.237579 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:3376: The name tf.log is deprecated. Please use tf.math.log instead.

W1013 13:40:55.245079 139621437822848 deprecation.py:323] From /usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/nn_impl.py:183: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


In [24]:
model.fit(X_train,y_train,batch_size=10,nb_epoch=100)

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  """Entry point for launching an IPython kernel.
W1013 13:41:05.225368 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:986: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.

W1013 13:41:05.293628 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:973: The name tf.assign is deprecated. Please use tf.compat.v1.assign instead.

W1013 13:41:05.351429 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:2741: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.

W1013 13:41:05.356005 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.p

Epoch 1/100


W1013 13:41:05.677031 139621437822848 module_wrapper.py:139] From /usr/local/lib/python2.7/dist-packages/keras/backend/tensorflow_backend.py:206: The name tf.variables_initializer is deprecated. Please use tf.compat.v1.variables_initializer instead.



7600/7600 [==============================] - 2s 212us/step - loss: 0.4990 - acc: 0.7941
Epoch 2/100
7600/7600 [==============================] - 1s 109us/step - loss: 0.4413 - acc: 0.7945
Epoch 3/100
7600/7600 [==============================] - 1s 109us/step - loss: 0.4402 - acc: 0.7945
Epoch 4/100
7600/7600 [==============================] - 1s 117us/step - loss: 0.4404 - acc: 0.7945
Epoch 5/100
7600/7600 [==============================] - 1s 117us/step - loss: 0.4346 - acc: 0.7945
Epoch 6/100
7600/7600 [==============================] - 1s 118us/step - loss: 0.4338 - acc: 0.7945
Epoch 7/100
7600/7600 [==============================] - 1s 116us/step - loss: 0.4301 - acc: 0.7945
Epoch 8/100
7600/7600 [==============================] - 1s 109us/step - loss: 0.4302 - acc: 0.8054
Epoch 9/100
7600/7600 [==============================] - 1s 109us/step - loss: 0.4289 - acc: 0.8192
Epoch 10/100
7600/7600 [==============================] - 1s 111us/step - loss: 0.4286 - acc: 0.8203
Epoch 11/10

In [0]:
#Predict Test Values
y_pred=model.predict(X_test)
y_pred=(y_pred>0.5)

In [26]:
y_pred

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

In [0]:
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)

In [28]:
cm

array([[1885,   40],
       [ 348,  127]])

In [29]:
accuracy=100*(1885+127)/(1885+40+348+127)
accuracy

83

## Review model

In [30]:
model.get_weights()

[array([[-7.68694431e-02, -9.96936336e-02, -9.89368558e-02,
         -1.39506370e-01, -1.02080807e-01, -4.46569383e-01],
        [-8.02881122e-02, -1.32445991e-02, -3.64191420e-02,
         -5.67941042e-03,  1.23557448e-01,  1.78192034e-02],
        [ 3.22585143e-02,  6.60263514e-03,  6.17739186e-02,
          1.29275527e-02, -2.68313079e-03,  1.93022483e-04],
        [ 1.19465195e-01,  1.25272110e-01,  1.32929236e-01,
          8.05454925e-02,  7.33737946e-02,  1.32638186e-01],
        [-6.79067850e-01, -5.98529935e-01, -6.24840260e-01,
         -6.07258558e-01,  6.61622643e-01, -4.11154538e-01],
        [-1.08001567e-02, -5.43737598e-02, -1.10540837e-02,
         -4.62289201e-03,  8.24535862e-02,  1.49225920e-01],
        [-9.93250981e-02, -9.00069326e-02, -7.64824897e-02,
         -1.48946131e-02,  1.35271382e-02, -1.12353660e-01],
        [ 3.16738635e-02,  3.18427868e-02,  4.73903790e-02,
          3.24411653e-02, -6.16444126e-02, -8.79073422e-03],
        [-3.38322259e-02,  4.545

### Run the model

# Model 1 withonly one layer accuracy is 29%

# Model 2 with Batch of 100 and learning rate of 0.01  increased Accuracy to  91.11%

# Model 3 with adding 3 hidden layers with sigmoid function and learning rate of 0.03 increased the accuracy to 91.15% 


In [0]:
model = keras.Sequential([
   
    keras.layers.Reshape((784,),input_shape=(28,28,)),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [0]:
model.fit(trainX, trainY, epochs=100,batch_size=100)

ValueError: ignored

In [31]:
model =Sequential()
model.add(Dense(output_dim=6,init="uniform",activation='relu',input_dim=11))
model.add(Dropout(rate=0.1))

model.add(Dense(output_dim=6,init="uniform",activation='relu',input_dim=11))
model.add(Dropout(rate=0.5))

model.add(Dense(output_dim=1,init="uniform",activation='sigmoid'))

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", activation="relu", input_dim=11, units=6)`
  
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", activation="relu", input_dim=11, units=6)`
  """
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:8: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(units=1, activation="sigmoid", kernel_initializer="uniform")`
  


In [0]:
#Compile the model - add Loss and Gradient Descent optimizer
model.compile(optimizer='adam', loss='binary_crossentropy',metrics=['accuracy'])

In [34]:
model.fit(X_train,y_train,batch_size=10,nb_epoch=100)

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  """Entry point for launching an IPython kernel.


Epoch 1/100
7600/7600 [==============================] - 1s 143us/step - loss: 0.5103 - acc: 0.7945
Epoch 2/100
7600/7600 [==============================] - 1s 109us/step - loss: 0.4553 - acc: 0.7945
Epoch 3/100
7600/7600 [==============================] - 1s 110us/step - loss: 0.4484 - acc: 0.7945
Epoch 4/100
7600/7600 [==============================] - 1s 111us/step - loss: 0.4516 - acc: 0.7945
Epoch 5/100
7600/7600 [==============================] - 1s 112us/step - loss: 0.4497 - acc: 0.7945
Epoch 6/100
7600/7600 [==============================] - 1s 111us/step - loss: 0.4485 - acc: 0.7945
Epoch 7/100
7600/7600 [==============================] - 1s 113us/step - loss: 0.4444 - acc: 0.7945
Epoch 8/100
7600/7600 [==============================] - 1s 112us/step - loss: 0.4490 - acc: 0.7945
Epoch 9/100
7600/7600 [==============================] - 1s 113us/step - loss: 0.4506 - acc: 0.7945
Epoch 10/100
7600/7600 [==============================] - 1s 114us/step - loss: 0.4505 - acc: 0.7945

In [38]:
model =Sequential()
model.add(Dense(output_dim=6,init="uniform",activation='relu',input_dim=11))
model.add(Dropout(rate=0.05))

model.add(Dense(output_dim=100,init="uniform",activation='relu'))
model.add(Dropout(rate=0.05))
model.add(Dense(output_dim=200,init="uniform",activation='relu'))
model.add(Dropout(rate=0.05))

model.add(Dense(output_dim=1,init="uniform",activation='sigmoid'))

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", activation="relu", input_dim=11, units=6)`
  
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(units=100, activation="relu", kernel_initializer="uniform")`
  """
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:7: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(units=200, activation="relu", kernel_initializer="uniform")`
  import sys
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:10: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(units=1, activation="sigmoid", kernel_initializer="uniform")`
  # Remove the CWD from sys.path while we load stuff.


In [0]:
#Compile the model - add Loss and Gradient Descent optimizer
model.compile(optimizer='adam', loss='binary_crossentropy',metrics=['accuracy'])

In [40]:
model.fit(X_train,y_train,batch_size=10,nb_epoch=100)

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  """Entry point for launching an IPython kernel.


Epoch 1/100
7600/7600 [==============================] - 2s 213us/step - loss: 0.4538 - acc: 0.8005
Epoch 2/100
7600/7600 [==============================] - 1s 148us/step - loss: 0.4189 - acc: 0.8284
Epoch 3/100
7600/7600 [==============================] - 1s 146us/step - loss: 0.4151 - acc: 0.8275
Epoch 4/100
7600/7600 [==============================] - 1s 146us/step - loss: 0.4141 - acc: 0.8305
Epoch 5/100
7600/7600 [==============================] - 1s 153us/step - loss: 0.4103 - acc: 0.8328
Epoch 6/100
7600/7600 [==============================] - 1s 151us/step - loss: 0.4111 - acc: 0.8308
Epoch 7/100
7600/7600 [==============================] - 1s 151us/step - loss: 0.4103 - acc: 0.8321
Epoch 8/100
7600/7600 [==============================] - 1s 151us/step - loss: 0.4086 - acc: 0.8333
Epoch 9/100
7600/7600 [==============================] - 1s 151us/step - loss: 0.4101 - acc: 0.8333
Epoch 10/100
7600/7600 [==============================] - 1s 148us/step - loss: 0.4105 - acc: 0.8328

In [0]:
#Predict Test Values
y_pred=model.predict(X_test)
y_pred=(y_pred>0.5)

In [42]:
cm=confusion_matrix(y_test,y_pred)
cm

array([[1827,   98],
       [ 225,  250]])

In [44]:
accuracy=100*(1827+250)/(1827+98+225+250)
accuracy

86

# Accuracy Improved to 86%
Added layers with 100 and 200 and learning rate of 0.05